### PR TITLE
[Enhencement](Nereids) add eliminate rollup

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
## Proposed changes

select c1, sum(c2) from t1 group by rollup(c1) having c1 > 1; ->
select c1, sum(c2) from t1 group by c1 having c1 > 1;

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        return ImmutableList.of(
+        logicalFilter(logicalAggregate(logicalProject(logicalRepeat()))).thenApply(ctx -> {
+            LogicalFilter filter = ctx.root;
+            LogicalAggregate agg = (LogicalAggregate) filter.child();
+            Set<Expression> groupNotNull = ExpressionUtils.inferNotNull(
+                    filter.getConjuncts(), agg.getOutputSet(), ctx.cascadesContext);
+            LogicalRepeat<Plan> repeat = (LogicalRepeat<Plan>) agg.child().child(0);
+            if (repeat.getGroupingSets().size() > 2) {
+                return null;
+            }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:50]: move this condition before infer not null

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java (hunk 1 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:49]: it make sense

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalRepeat;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Set;
+
+/** EliminateRollup */
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:37]: add simple case for this rule

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java [hunk 3]
```diff
@@ -204,6 +205,7 @@ public class Rewriter extends AbstractBatchJobExecutor {
                             new EliminateLimit(),
                             new EliminateFilter(),
                             new EliminateAggregate(),
+                            new EliminateRollup(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java:208]: it is better work togather with push filter

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+public class EliminateRollup implements RewriteRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+        logicalFilter(logicalAggregate(logicalProject(logicalRepeat()))).thenApply(ctx -> {
+            LogicalFilter filter = ctx.root;
+            LogicalAggregate agg = (LogicalAggregate) filter.child();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:44]: use generic

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            // slots in rollup should be null reject
+            if (!ExpressionUtils.getInputSlotSet(groupNotNull).contains(groupingExpr)) {
+                return null;
+            }
+
+            ImmutableList.Builder<NamedExpression> newOutput = ImmutableList.builder();
+            for (int i = 0; i < agg.getOutputExpressions().size(); i++) {
+                NamedExpression expr = (NamedExpression) agg.getOutputExpressions().get(i);
+                if (expr instanceof VirtualSlotReference
+                        && ((VirtualSlotReference) expr).getName().equals("GROUPING_ID")) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:62]: use Repeat.COL_GROUPING_ID

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            LogicalRepeat<Plan> repeat = (LogicalRepeat<Plan>) agg.child().child(0);
+            // check rollup only have one parameter, which grouping sets would be 2
+            if (repeat.getGroupingSets().size() != 2 || !repeat.getGroupingSets().get(1).isEmpty()) {
+                return null;
+            }
+            Set<Expression> groupNotNull = ExpressionUtils.inferNotNull(
+                    filter.getConjuncts(), agg.getOutputSet(), ctx.cascadesContext);
+            Expression groupingExpr = repeat.getGroupingSets().get(0).get(0);
+            // slots in rollup should be null reject
+            if (!ExpressionUtils.getInputSlotSet(groupNotNull).contains(groupingExpr)) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:54]: should check project only has Slot or Alias(Slot). if project has Alias(Slot) u should do replace for expressions in aggregate

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                    continue;
+                } else {
+                    newGroupByExprs.add(expr);
+                }
+            }
+            // eliminate repeat
+            return new LogicalFilter(filter.getConjuncts(),
+                    new LogicalAggregate<>((List<Expression>) newGroupByExprs.build(),
+                    (List<NamedExpression>) newOutput.build(), (Plan) agg.child().child(0).child(0)));
+        }).toRule(RuleType.ELIMINATE_AGGREGATE),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:83]: use new rule type

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        );
+    }
+
+    private boolean isSame(List<Expression> list1, List<Expression> list2) {
+        return list1.size() == list2.size() && list2.containsAll(list1);
+    }
+
+    private boolean onlyHasSlots(List<? extends Expression> exprs) {
+        return exprs.stream().allMatch(SlotReference.class::isInstance);
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:129]: remove unuse functions

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            LogicalFilter filter = ctx.root;
+            LogicalAggregate agg = (LogicalAggregate) filter.child();
+            LogicalRepeat<Plan> repeat = (LogicalRepeat<Plan>) agg.child().child(0);
+            // check rollup only have one parameter, which grouping sets would be 2
+            if (repeat.getGroupingSets().size() != 2 || !repeat.getGroupingSets().get(1).isEmpty()) {
+                return null;
+            }
+            Set<Expression> groupNotNull = ExpressionUtils.inferNotNull(
+                    filter.getConjuncts(), agg.getOutputSet(), ctx.cascadesContext);
+            Expression groupingExpr = repeat.getGroupingSets().get(0).get(0);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:52]: why only process the first group by key? what about `grouping sets(a,b,c)`?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                    (List<NamedExpression>) newOutput.build(), (Plan) agg.child().child(0).child(0)));
+        }).toRule(RuleType.ELIMINATE_AGGREGATE),
+        logicalAggregate(logicalProject(logicalRepeat())).thenApply(ctx -> {
+            LogicalAggregate agg = ctx.root;
+            LogicalRepeat<Plan> repeat = (LogicalRepeat<Plan>) agg.child().child(0);
+            if (repeat.getGroupingSets().size() != 2 || !repeat.getGroupingSets().get(1).isEmpty()) {
+                return null;
+            }
+            Expression groupingExpr = repeat.getGroupingSets().get(0).get(0);
+            if (!(groupingExpr.isSlot() && !((SlotReference) groupingExpr).nullable())) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateRollup.java:91]: in which case `groupingExpr.nullable() == false`?

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Pull request closed without merging**